src/Language/Futhark/Interpreter/AD.hs

```diff
@@ -0,0 +1,289 @@
+{-# LANGUAGE LambdaCase #-}
+module Language.Futhark.Interpreter.AD
+  (
+    ADSeed (..),
+    ADValue (..),
+    Op (..),
+    Tape (..),
+
+    value,
+    depth,
+    primal,
+
+    deriveVjp,
+
+    valueAsType,
+
+    addFor,
+
+    doOp,
+    handleOp,
+  )
+where
+
+import Language.Futhark.Primitive
+import Data.List
+import Data.Map qualified as M
+import Data.Set qualified as S
+
+import Futhark.AD.Derivatives
+import Futhark.Analysis.PrimExp
+import Language.Futhark.Core
+
+import Debug.Trace qualified as DBG
+import Control.Monad (zipWithM)
+import Data.Maybe (fromJust)
+import Data.Foldable (foldlM)
+
+data ADSeed
```

Comment:
Add a comment remarking on what the meaning of "seed" is (both here and in your implementation in general).